group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilderTest.java

```diff
@@ -0,0 +1,688 @@
+
+        public TargetAssignmentBuilderTestContext addGroupMember(
+            String memberId,
+            List<String> subscriptions,
+            Map<Uuid, Set<Integer>> targetPartitions
+        ) {
+            members.put(memberId, new ConsumerGroupMember.Builder(memberId)
+                .setSubscribedTopicNames(subscriptions)
+                .setRebalanceTimeoutMs(5000)
+                .build());
+
+            targetAssignments.put(memberId, new Assignment(
+                (byte) 0,
+                targetPartitions,
+                VersionedMetadata.EMPTY
+            ));
+
+            return this;
+        }
+
+        public TargetAssignmentBuilderTestContext addTopicMetadata(
+            Uuid topicId,
+            String topicName,
+            int numPartitions
+        ) {
+            subscriptionMetadata.put(topicName, new TopicMetadata(
+                topicId,
+                topicName,
+                numPartitions
+            ));
+            return this;
+        }
+
+        public TargetAssignmentBuilderTestContext updateMemberSubscription(
+            String memberId,
+            List<String> subscriptions
+        ) {
+            return updateMemberSubscription(
+                memberId,
+                subscriptions,
+                Optional.empty(),
+                Optional.empty()
+            );
+        }
+
+        public TargetAssignmentBuilderTestContext updateMemberSubscription(
+            String memberId,
+            List<String> subscriptions,
+            Optional<String> instanceId,
+            Optional<String> rackId
+        ) {
+            ConsumerGroupMember existingMember = members.get(memberId);
+            ConsumerGroupMember.Builder builder;
+            if (existingMember != null) {
+                builder = new ConsumerGroupMember.Builder(existingMember);
+            } else {
+                builder = new ConsumerGroupMember.Builder(memberId);
+            }
+            updatedMembers.put(memberId, builder
+                .setSubscribedTopicNames(subscriptions)
+                .maybeUpdateInstanceId(instanceId)
+                .maybeUpdateRackId(rackId)
+                .build());
+            return this;
+        }
+
+        public TargetAssignmentBuilderTestContext removeMemberSubscription(
+            String memberId
+        ) {
+            this.updatedMembers.put(memberId, null);
+            return this;
+        }
+
+        public TargetAssignmentBuilderTestContext prepareMemberAssignment(
+            String memberId,
+            Map<Uuid, Set<Integer>> assignment
+        ) {
+            assignments.put(memberId, new org.apache.kafka.coordinator.group.assignor.MemberAssignment(assignment));
+            return this;
+        }
+
+        private AssignmentMemberSpec assignmentMemberSpec(
+            ConsumerGroupMember member,
+            Assignment targetAssignment
+        ) {
+            Set<Uuid> subscribedTopics = new HashSet<>();
```

Comment:
This seems to be mostly the same code as addMemberSpec. I'm wondering if there is a way to avoid code duplication here.

Reply:
Good call. Let me pull that code into an helper method.